File: src/SOP8/Utility/Response/Resource/MessageMap.cs
using System.Collections.Generic;

namespace Response.Resource
{
    public class MessageMap
    {
        protected Dictionary<string, object> m_maps = null;
        protected string m_strTarget = "";

        protected MessageMap(Dictionary<string, object> messageMaps)
        {
            m_maps = messageMaps;
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTag, string strMessage)
        {
            Dictionary<string, object> dicMessageMaps;

            if (messageMaps.TryGetValue(type, out dicMessageMaps) == false)
            {
                dicMessageMaps = new Dictionary<string, object>();
                messageMaps[type] = dicMessageMaps;
            }

            dicMessageMaps[strTag] = strMessage;
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strMessage)
        {
            Dictionary<string, object> dicRoot = GetRootMap(messageMaps, type, strTagLevel1);
            SetMessage(dicRoot, strTagLevel2, strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);

            int nCount = tags.Count;

            for (int i=0;i<nCount-1;i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount-1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strTagLevel5, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);
            tags.Add(strTagLevel5);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strTagLevel5, string strTagLevel6, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);
            tags.Add(strTagLevel5);
            tags.Add(strTagLevel6);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strTagLevel5, string strTagLevel6, string strTagLevel7, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);
            tags.Add(strTagLevel5);
            tags.Add(strTagLevel6);
            tags.Add(strTagLevel7);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strTagLevel5, string strTagLevel6, string strTagLevel7, string strTagLevel8, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);
            tags.Add(strTagLevel5);
            tags.Add(strTagLevel6);
            tags.Add(strTagLevel7);
            tags.Add(strTagLevel8);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strTagLevel5, string strTagLevel6, string strTagLevel7, string strTagLevel8, string strTagLevel9, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);
            tags.Add(strTagLevel5);
            tags.Add(strTagLevel6);
            tags.Add(strTagLevel7);
            tags.Add(strTagLevel8);
            tags.Add(strTagLevel9);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTagLevel1, string strTagLevel2, string strTagLevel3, string strTagLevel4, string strTagLevel5, string strTagLevel6, string strTagLevel7, string strTagLevel8, string strTagLevel9, string strTagLevel10, string strMessage)
        {
            Dictionary<string, object> dicParent = GetRootMap(messageMaps, type, strTagLevel1);

            List<string> tags = new List<string>();
            tags.Add(strTagLevel2);
            tags.Add(strTagLevel3);
            tags.Add(strTagLevel4);
            tags.Add(strTagLevel5);
            tags.Add(strTagLevel6);
            tags.Add(strTagLevel7);
            tags.Add(strTagLevel8);
            tags.Add(strTagLevel9);
            tags.Add(strTagLevel10);

            int nCount = tags.Count;

            for (int i = 0; i < nCount - 1; i++)
            {
                string strTag = tags[i];
                dicParent = GetMap(dicParent, strTag);
            }

            SetMessage(dicParent, tags[nCount - 1], strMessage);
        }

        protected static void SetMessage(Dictionary<string, object> messageMaps, string strTag, string strMessage)
        {
            messageMaps[strTag] = strMessage;
        }

        protected static Dictionary<string, object> GetRootMap(Dictionary<ID.LanguageTypes, Dictionary<string, object>> messageMaps, ID.LanguageTypes type, string strTag)
        {
            Dictionary<string, object> dicMessageMaps;

            if (messageMaps.TryGetValue(type, out dicMessageMaps) == false)
            {
                dicMessageMaps = new Dictionary<string, object>();
                messageMaps[type] = dicMessageMaps;
            }

            object dicSubMessageMaps;

            if (dicMessageMaps.TryGetValue(strTag, out dicSubMessageMaps) == false)
            {
                dicSubMessageMaps = new Dictionary<string, object>();
                dicMessageMaps[strTag] = dicSubMessageMaps;
            }

            return (Dictionary<string, object>)dicSubMessageMaps;
        }

        protected static Dictionary<string, object> GetMap(Dictionary<string, object> parentMap, string strTag)
        {
            object dicSubMessageMaps;

            if (parentMap.TryGetValue(strTag, out dicSubMessageMaps) == false)
            {
                dicSubMessageMaps = new Dictionary<string, object>();
                parentMap[strTag] = dicSubMessageMaps;
            }

            return (Dictionary<string, object>)dicSubMessageMaps;
        }

        public MessageMap Get(string strTag)
        {
            if (m_maps == null)
                return this;

            object dicMessageMaps = null;

            if (m_maps.TryGetValue(strTag, out dicMessageMaps))
            {
                if (dicMessageMaps is string)
                {
                    m_strTarget = (string)dicMessageMaps;
                    m_maps = null;
                    return this;
                }

                m_maps = (Dictionary<string, object>)dicMessageMaps;
                return this;
            }

            m_maps = null;
            return this;
        }

        public string Value(string strTag)
        {
            if (m_maps == null)
                return m_strTarget;

            object target;

            if (m_maps.TryGetValue(strTag, out target))
            {
                if (target is string)
                {
                    m_maps = null;
                    m_strTarget = (string)target;
                    return m_strTarget;
                }
            }

            return "";
        }

        public string Value()
        {
            return m_strTarget;
        }

        public virtual MessageMap Get(ID.LanguageTypes type)
        {
            return null;
        }
    }
}

File: src/SOP8/Egovframework/Base/base-utility/src/main/java/egovframework/base/response/resource/MessageMap.java
package egovframework.base.response.resource;

import java.util.*;
import egovframework.base.response.resource.ID;

public class MessageMap {
    protected Map<String, Object> maps = null;
    protected String target = "";

    protected MessageMap(Map<String, Object> messageMaps) {
        this.maps = messageMaps;
    }

    protected static void setMessage(Map<ID.LanguageTypes, Map<String, Object>> messageMaps, ID.LanguageTypes type, String strTag, String strMessage) {
        Map<String, Object> dicMessageMaps = messageMaps.get(type);

        if (dicMessageMaps == null) {
            dicMessageMaps = new HashMap<>();
            messageMaps.put(type, dicMessageMaps);
        }

        dicMessageMaps.put(strTag, strMessage);
    }

    protected static void setMessage(Map<ID.LanguageTypes, Map<String, Object>> messageMaps, ID.LanguageTypes type, String strTagLevel1, String strTagLevel2, String strMessage) {
        Map<String, Object> dicRoot = getRootMap(messageMaps, type, strTagLevel1);
        setMessage(dicRoot, strTagLevel2, strMessage);
    }

    // Generic method for multiple levels
    protected static void setMessage(Map<ID.LanguageTypes, Map<String, Object>> messageMaps, ID.LanguageTypes type, String strTagLevel1, String... tagsAndMessage) {
        Map<String, Object> dicParent = getRootMap(messageMaps, type, strTagLevel1);

        List<String> tags = new ArrayList<>(Arrays.asList(tagsAndMessage));
        String strMessage = tags.remove(tags.size() - 1); // Last element is message

        for (int i = 0; i < tags.size() - 1; i++) {
            String strTag = tags.get(i);
            dicParent = getMap(dicParent, strTag);
        }

        setMessage(dicParent, tags.get(tags.size() - 1), strMessage);
    }

    protected static void setMessage(Map<String, Object> messageMaps, String strTag, String strMessage) {
        messageMaps.put(strTag, strMessage);
    }

    protected static Map<String, Object> getRootMap(Map<ID.LanguageTypes, Map<String, Object>> messageMaps, ID.LanguageTypes type, String strTag) {
        Map<String, Object> dicMessageMaps = messageMaps.get(type);

        if (dicMessageMaps == null) {
            dicMessageMaps = new HashMap<>();
            messageMaps.put(type, dicMessageMaps);
        }

        Object dicSubMessageMaps = dicMessageMaps.get(strTag);

        if (dicSubMessageMaps == null) {
            dicSubMessageMaps = new HashMap<String, Object>();
            dicMessageMaps.put(strTag, dicSubMessageMaps);
        }

        return (Map<String, Object>) dicSubMessageMaps;
    }

    protected static Map<String, Object> getMap(Map<String, Object> parentMap, String strTag) {
        Object dicSubMessageMaps = parentMap.get(strTag);

        if (dicSubMessageMaps == null) {
            dicSubMessageMaps = new HashMap<String, Object>();
            parentMap.put(strTag, dicSubMessageMaps);
        }

        return (Map<String, Object>) dicSubMessageMaps;
    }

    public MessageMap get(String strTag) {
        if (maps == null)
            return this;

        Object dicMessageMaps = maps.get(strTag);

        if (dicMessageMaps != null) {
            if (dicMessageMaps instanceof String) {
                target = (String) dicMessageMaps;
                maps = null;
                return this;
            }

            maps = (Map<String, Object>) dicMessageMaps;
            return this;
        }

        maps = null;
        return this;
    }

    public String value(String strTag) {
        if (maps == null)
            return target;

        Object targetObj = maps.get(strTag);

        if (targetObj instanceof String) {
            maps = null;
            target = (String) targetObj;
            return target;
        }

        return "";
    }

    public String value() {
        return target;
    }

    public MessageMap get(ID.LanguageTypes type) {
        return null;
    }
}